pkg/virtctl/adm/logverbosity/logverbosity.go

```diff
@@ -0,0 +1,447 @@
+	// but we have to manage which components have the existing verbosity, which makes the code complicated
+	setVerbosity(lvMap, options, &patchData, &op, &path)
+
+	return json.Marshal(patchData)
+}
+
+func findOperation(cmd *cobra.Command, options map[string]uint) (operation, error) {
+	isShow, isSet := false, false
+
+	for component := virtAPI; component <= all; component++ {
+		componentName := getComponentNameByVirtComponent(component)
+
+		// check if the flag for the component is specified
+		if !cmd.Flags().Changed(componentName) {
+			continue // do nothing for the component
+		}
+
+		// if flag is specified, it means either set or show
+		// if the value = noArg, it means show
+		// if the value != noArg, it means set
+		isShow = isShow || verbosities[component] == noArg
+		isSet = isSet || verbosities[component] != noArg
+
+		// check whether the verbosity is in the range
+		if verbosities[component] != noArg && verbosities[component] > maxVerbosity {
+			return nop, fmt.Errorf("%s: log verbosity must be %d-%d", componentName, minVerbosity, maxVerbosity)
+		}
+
+		// add a componentName and its verbosity to the map
+		// in case of show: verbosity = noArg
+		// in case of set: verbosity = 0-9
+		// in case of reset: no information added to the map
+		options[componentName] = verbosities[component]
+	}
+
+	// do not distinguish between set and reset at this point
+	// because set and reset can coexist
+	if isReset {
+		isSet = true
+	}
+
+	if isShow && isSet {
+		return nop, fmt.Errorf("only show or set is allowed")
+	}
+
+	if isShow {
+		return show, nil
+	} else if isSet {
+		return set, nil
+	}
+
+	return nop, nil
+}
+
+func (c *Command) RunE(cmd *cobra.Command) error {
+	// get client
+	virtClient, err := kubecli.GetKubevirtClientFromClientConfig(c.clientConfig)
+	if err != nil {
+		return err
+	}
+	// get install namespace and name
+	namespace, name, err := detectInstallNamespaceAndName(virtClient)
+	if err != nil {
+		return err
+	}
+	// get KubeVirt CR
+	kv, err := virtClient.KubeVirt(namespace).Get(name, &k8smetav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	// check the operation type (nop/show/set), and set the options map to use the map for show and set operations
+	options := map[string]uint{} // key: component name, value: verbosity
+	op, err := findOperation(cmd, options)
+	if err != nil {
+		return err
+	}
+
+	switch op {
+	case nop:
+		err := cmd.Help()
+		if err != nil {
+			return err
+		}
+		return errors.New("no flag specified - expecting at least one flag")
+	case show:
+		messages, err := createShowMessage(kv, options)
+		if err != nil {
+			return err
+		}
+		for _, message := range messages {
+			cmd.Println(message)
+		}
+	case set: // set and/or reset
+		// create patch data
+		patchData, err := createPatch(kv, options)
+		if err != nil {
+			return err
+		}
+		// apply patch, if patch data exists
+		if len(patchData) != 0 {
```

Comment:
This is not enough, I think. Quick trying the following
```Go
arr := []string{}
b, _ := json.Marshal(&arr)
fmt.Printf("%d %s\n", len(b), string(b))
```
Outputs:
```
2 []
```

I'd say is best to not have to check this at all here. Just return an error if we want to `createPatch()` but it does not have any data to do so.

Reply:
Great catch! Many thanks.

I cared about the reset only (= no set operation after reset) and the empty `currentLV` (= no existing verbosity in the KubeVirt CR) case. In this case, `patchData = {}`. And, I tried to avoid an unnecessary `patch` function call. However, the overhead of this unnecessary `patch` call should not be a big deal. And even in this case, the `patch` function will not return an error. So I agree and I'll remove the check.

Besides, the `len` is not 0, so it makes no sense to check it here.
(I do not know why it is not 0, though.)